### PR TITLE
net: Support packet filter (firewall) and corresponding iptables/ip6tables

### PR DESCRIPTION
**Pull request by wengzhe**
## Summary

Patches included:
- net: Support IP packet filter
- net/netfilter: Add filter table in iptables.
- net: Add set/getsockopt options compatible with ip6tables
- net/netfilter: Add filter table in ip6tables
- Documentation: Add docs for ipfilter

Add a firewall compatible with Linux's iptables and ip6tables, with chains at similar points in the packet processing path. More details and examples are in the doc.

```
   NIC ──> ipv[46]_input ─┬─> ipv[46]_forward ──> [FORWARD CHAIN] ──> devif_poll_out ──> NIC
                          │                                                 ^
                          │                  ┌─>  tcp  ─┐                   │
                          │                  ├─>  udp  ─┤                   │
                          └─> [INPUT CHAIN] ─┼─> icmp  ─┼─> [OUTPUT CHAIN] ─┘
                                             ├─> icmp6 ─┤
                                             └─>  ...  ─┘
```

## Impact

Changes limited in `net/ipfilter`(firewall) `net/netfilter` (interface) and minimized in other modules, all controlled by `CONFIG_NET_IPFILTER`

## Testing

Together with https://github.com/apache/nuttx-apps/pull/2415

### CODE REVIEW

File: net/ipfilter/ipfilter.c [hunk 1]
```diff
@@ -0,0 +1,632 @@
...
+ * Input Parameters:
+ *   family - The address family of the filter entry
+ *
+ * Returned Value:
+ *   A pointer to the newly allocated filter entry.  NULL is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+FAR struct ipfilter_entry_s *ipfilter_cfg_alloc(sa_family_t family)
```
Comment (anchao) [inline net/ipfilter/ipfilter.c:367]: remove this method if nobody using

File: net/netfilter/ip6t_sockopt.c [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+  int ret;
+  FAR struct ip6t_replace *new_repl;
+  FAR struct ip6t_table_s *table = ip6t_table(repl->name);
+
+  if (table == NULL || table->repl == NULL)
+    {
+      return -ENOENT;
+    }
+
+  if (len < sizeof(*repl) || len != sizeof(*repl) + repl->size ||
```
Comment (anchao) [inline net/netfilter/ip6t_sockopt.c:281]:
```suggestion
  if (len != sizeof(*repl) + repl->size ||
```

File: net/netfilter/ipt_filter.c [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+
+#ifdef CONFIG_NET_IPv6
+static FAR struct ipv6_filter_entry_s *
+convert_ipv6entry(FAR const struct ip6t_entry *entry)
+{
+  FAR const struct xt_entry_match *match;
+  FAR const struct xt_entry_target *target;
+  FAR struct ipv6_filter_entry_s *filter;
+
+  filter = kmm_zalloc(sizeof(struct ipv6_filter_entry_s));
```
Comment (anchao) [inline net/netfilter/ipt_filter.c:282]: ditto

File: net/netfilter/ip6t_sockopt.c [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+
+#include <nuttx/kmalloc.h>
+
+#include "netfilter/iptables.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define SWAP(a,b,t)    do { (t) = (a); (a) = (b); (b) = (t); } while (0)
```
Comment (anchao) [inline net/netfilter/ip6t_sockopt.c:40]:
```suggestion
#define SWAP(a,b,t)    do { FAR void *t; (t) = (a); (a) = (b); (b) = (t); } while (0)
```

File: net/devif/devif_poll.c [hunk 5]
```diff
@@ -184,6 +185,32 @@ static void devif_packet_conversion(FAR struct net_driver_s *dev,
...
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_NET_IPFILTER
+static int devif_poll_local_out(FAR struct net_driver_s *dev,
+                                devif_poll_callback_t callback)
+{
+  /* Maybe we need to reply REJECT to ourself, so filter before loopback. */
+
+  ipfilter_out(dev);
```
[the remark this answers is not in the document]
Comment (wengzhe) [reply net/devif/devif_poll.c:207]: Do you mean moving `CONFIG_NET_IPFILTER` from L201 to L206 and always keep `devif_poll_local_out`?

File: net/netfilter/ip6t_sockopt.c (hunk 4 above)
Comment (wengzhe) [reply net/netfilter/ip6t_sockopt.c:40]: @anchao Do you mean to change `SWAP(a,b,t)` to `SWAP(a,b)`?

File: net/netfilter/ip6t_sockopt.c (hunk 2 above)
Comment (wengzhe) [reply net/netfilter/ip6t_sockopt.c:281]: If `len < sizeof(*repl)`, it's dangerous to access `repl->size`, so I check `len < sizeof(*repl)` first.

File: net/ipfilter/ipfilter.c (hunk 1 above)
Comment (wengzhe) [reply net/ipfilter/ipfilter.c:367]: Thanks! I designed this for use in other places but forgot it. Now `convert_ipv4entry` and `convert_ipv6entry` will use it and we may optimize the allocation logic in this function later.

File: net/netfilter/ipt_filter.c (hunk 3 above)
Comment (wengzhe) [reply net/netfilter/ipt_filter.c]: Done.

File: net/netfilter/ipt_filter.c [hunk 6]
```diff
@@ -0,0 +1,593 @@
...
+
+#ifdef CONFIG_NET_IPv4
+static FAR struct ipv4_filter_entry_s *
+convert_ipv4entry(FAR const struct ipt_entry *entry)
+{
+  FAR const struct xt_entry_match *match;
+  FAR const struct xt_entry_target *target;
+  FAR struct ipv4_filter_entry_s *filter;
+
+  filter = kmm_zalloc(sizeof(struct ipv4_filter_entry_s));
```
[the remark this answers is not in the document]
Comment (wengzhe) [reply net/netfilter/ipt_filter.c]: Done.

File: net/netfilter/ip6t_sockopt.c (hunk 4 above)
Comment (anchao) [reply net/netfilter/ip6t_sockopt.c:40]: Yep, this private macro only for pointer objects, you could keep the current style if which have other references

File: net/devif/devif_poll.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (anchao) [reply net/devif/devif_poll.c:207]: Yes, would it be easier to read if we only protected the call around ipfilter_out(), or there is something I missed?

[the remark this answers is not in the document]
Comment (wengzhe) [reply net/devif/devif_poll.c:207]: Done.

File: net/netfilter/ip6t_sockopt.c (hunk 4 above)
Comment (wengzhe) [reply net/netfilter/ip6t_sockopt.c]: Done.

**Review by anchao — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Comment by jerpelea** [comment]
please add commit messages

File: net/devif/devif_poll.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (anchao) [reply net/devif/devif_poll.c:207]: Sorry, please ignore my comments which will caused compiler warning, the previous implement is better.

[the remark this answers is not in the document]
Comment (wengzhe) [reply net/devif/devif_poll.c:207]: > Yes, would it be easier to read if we only protected the call around ipfilter_out(), or there is something I missed?

@anchao OK, now I found the case that I don't like, it's just an error we met in CI: `error: 'devif_poll_local_out' defined but not used [-Werror=unused-function]`
Then I'll need to write a lot of checks outside `devif_poll_local_out` like `#if defined(NET_TCP_HAVE_STACK) || defined(NET_UDP_HAVE_STACK) || (defined(CONFIG_NET_ICMP) && defined(CONFIG_NET_ICMP_SOCKET)) || defined(CONFIG_NET_ICMPv6_SOCKET) || defined(CONFIG_NET_ICMPv6_NEIGHBOR)` which I think is a little bit heavy.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**